libnopegl/src/glsl/helper_noise.glsl

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright 2023 Matthieu Bouron <matthieu.bouron@gmail.com>
+ * Copyright 2023 Clément Bœsch <u pkh.me>
+ * Copyright 2023 Nope Foundry
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include helper_misc_utils.glsl
+
+float u32tof32(highp uint x)
+{
+    uint y = (0x7FU << 23U) | (x >> 9U);
+    return (uintBitsToFloat(y) - 1.0);
+}
+
+highp uint lowbias32(highp uint x)
+{
+    x ^= x >> 16;
+    x *= 0x7feb352dU;
+    x ^= x >> 15;
+    x *= 0x846ca68bU;
+    x ^= x >> 16;
+    return x;
+}
+
+uint hash(uvec2 x)
+{
+    return lowbias32(x.x ^ lowbias32(x.y));
+}
+
+uint hash(uvec3 x)
+{
+    return lowbias32(x.x ^ lowbias32(x.y ^ lowbias32(x.z)));
+}
+
+float random(uvec3 x)
+{
+    return u32tof32(hash(x));
+}
+
+float random_grad(uvec2 x, vec2 f)
+{
+    float r = u32tof32(hash(x)) * 2.0 * ngli_pi;
+    return dot(vec2(cos(r), sin(r)), f);
+}
+
+float random_grad(uvec3 x, vec3 f)
+{
+    uint hash = hash(x) & 15U;
+    float u = hash < 8U ? f.x : f.y;
+    float vt = ((hash == 12U) || (hash == 14U)) ? f.x : f.z;
+    float v = hash < 4U ? f.y : vt;
+    return ((hash & 1U) != 0U ? -u : u) + ((hash & 2U) != 0U ? -v : v);
+}
+
+float noise_perlin(vec2 t, uint seed)
+{
+    vec2 i = floor(t);
+    vec2 f = fract(t);
+    uvec2 x = uvec2(i) + seed;
+
+    float y0 = random_grad(x + uvec2(0, 0), f - vec2(0, 0));
+    float y1 = random_grad(x + uvec2(1, 0), f - vec2(1, 0));
+    float y2 = random_grad(x + uvec2(0, 1), f - vec2(0, 1));
+    float y3 = random_grad(x + uvec2(1, 1), f - vec2(1, 1));
+
+    vec2 a = f * f * f * (f * (f * 6.0 - 15.0) + 10.0);
+    return(mix(mix(y0, y1, a.x), mix(y2, y3, a.x), a.y));
+}
+
+float noise_perlin(vec3 t, uint seed)
+{
+    vec3 i = floor(t);
+    vec3 f = fract(t);
+    uvec3 x = uvec3(i) + seed;
```

Comment:
I'm assuming the seed doesn't need to be hashed? I know we haven't done that in the noise node, but I'm starting to question it.

Reply:
It appears that ±1 completely shuffle the output so I guess that's fine (not exactly sure why though)